src/via.c

```diff
@@ -331,28 +340,134 @@ via1_irq()
 // for now, just assume that all user ports are not connected
 // and reads return output register (open bus behavior)
 
+
+static bool attempt_peripheral_load = true;
+static LIBRARY_TYPE user_peripheral_dl = NULL;
+static user_port_init_t user_port_init = NULL;
+
+static user_port_t user_port;
+
 void
-via2_init()
+via2_init(char const *user_peripheral_path)
 {
 	via_init(&via[1]);
+	if (attempt_peripheral_load && user_peripheral_path) {
+		attempt_peripheral_load = false;
+		user_peripheral_dl = LOAD_LIBRARY(user_peripheral_path);
+		if (user_peripheral_dl) {
+			user_port_init = GET_FUNCTION(user_peripheral_dl, "x16_user_port_init");
+		}
+		if (user_peripheral_dl == NULL || user_port_init == NULL) {
+			fprintf(stderr, "failed to load user peripheral %s:\n\t%s\n",
+			       user_peripheral_path, LIBRARY_ERROR());
+			fprintf(stderr, "continuing with empty user port.\n");
+			if (user_peripheral_dl) CLOSE_LIBRARY(user_peripheral_dl);
+		}
+	}
+	// TODO Do we really want to reset the user peripherals every time?
+	if (user_port_init) {
+		bool error = false;
+		if (user_port_init(&user_port) < 0) {
+			fprintf(stderr, "error initializing user peripheral\n");
+			error = true;
```

Comment:
This should return a proper error message. e.g. via `const char *user_port_error()` that is valid until the next call.

Reply:
Do you mean that the peripheral should retain its error, or the via code should retain userport errors, or both?